rust/src/qfunction.rs

```diff
@@ -0,0 +1,532 @@
+impl fmt::Display for QFunctionCore {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let mut ptr = std::ptr::null_mut();
+        let mut sizeloc = crate::MAX_BUFFER_LENGTH;
+        let cstring = unsafe {
+            let file = bind_ceed::open_memstream(&mut ptr, &mut sizeloc);
+            bind_ceed::CeedQFunctionView(self.ptr, file);
+            bind_ceed::fclose(file);
+            CString::from_raw(ptr)
+        };
+        cstring.to_string_lossy().fmt(f)
+    }
+}
+/// View a QFunction
+///
+/// ```
+/// # use libceed::prelude::*;
+/// # let ceed = libceed::Ceed::default_init();
+/// let mut user_f = |
+///   q: usize,
+///   inputs: &[&[f64]],
+///   outputs: &mut [&mut [f64]],
+/// | -> i32
+/// {
+///   // Inputs
+///   let u = &inputs[0];
+///   let weights = &inputs[1];
+///   // Outputs
+///   let v = &mut outputs[0];
+///
+///   // Loop over quadrature points
+///   v.iter_mut()
+///     .zip(u.iter().zip(weights.iter()))
+///     .for_each(|(v, (u, w))| *v = u * w);
+///
+///   // Return clean error code
+///   0
+/// };
+///
+/// let mut qf = ceed.q_function_interior(1, Box::new(user_f));
+///
+/// qf.add_input("u", 1, EvalMode::Interp);
+/// qf.add_input("weights", 1, EvalMode::Weight);
+///
+/// qf.add_output("v", 1, EvalMode::Interp);
+///
+/// println!("{}", qf);
+/// ```
+impl fmt::Display for QFunction {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        self.qf_core.fmt(f)
+    }
+}
+
+/// View a QFunction by Name
+///
+/// ```
+/// # use libceed::prelude::*;
+/// # let ceed = libceed::Ceed::default_init();
+/// let qf = ceed.q_function_interior_by_name("Mass1DBuild");
+/// println!("{}", qf);
+/// ```
+impl fmt::Display for QFunctionByName {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        self.qf_core.fmt(f)
+    }
+}
+
+// -----------------------------------------------------------------------------
+// Core functionality
+// -----------------------------------------------------------------------------
+impl QFunctionCore {
+    // Common implementation
+    pub fn apply(&self, Q: i32, u: &[Vector], v: &[Vector]) {
+        let mut u_c = [std::ptr::null_mut(); crate::MAX_QFUNCTION_FIELDS];
+        for i in 0..std::cmp::min(crate::MAX_QFUNCTION_FIELDS, u.len()) {
+            u_c[i] = u[i].ptr;
+        }
+        let mut v_c = [std::ptr::null_mut(); crate::MAX_QFUNCTION_FIELDS];
+        for i in 0..std::cmp::min(crate::MAX_QFUNCTION_FIELDS, v.len()) {
+            v_c[i] = v[i].ptr;
+        }
+        unsafe {
+            bind_ceed::CeedQFunctionApply(self.ptr, Q, u_c.as_mut_ptr(), v_c.as_mut_ptr());
+        }
+    }
+}
+
+// -----------------------------------------------------------------------------
+// User QFunction Closure
+// -----------------------------------------------------------------------------
+pub type QFunctionUserClosure = dyn FnMut(usize, &[&[f64]], &mut [&mut [f64]]) -> i32;
```

Comment:
I'm wondering if we could instead directly have arrays of slices of size `MAX_QFUNCTION_FIELDS` instead of slices of slices, could that prevent having to copy all the pointers in `apply`?

Reply:
We would still need to make slices from the raw pointers, since I'd like the users not to have to do `unsafe` things and using raw pointers is `unsafe`